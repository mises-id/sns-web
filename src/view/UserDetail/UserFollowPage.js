/*
 * @Author: lmk
 * @Date: 2022-01-05 15:51:32
 * @LastEditTime: 2022-01-12 19:09:09
 * @LastEditors: lmk
 * @Description: user following and followers
 */
import Navbar from '@/components/NavBar';
import { useRouteState } from '@/utils';
import React, { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Panel, Tabs } from 'zarm';
import Following from '../Following';
import "./index.scss";
const UserFollowPage = (props)=>{
  const [pageTitle, setpageTitle] = useState('')
  const {t} = useTranslation()
  const [value, setvalue] = useState(0)
  const [uid, setuid] = useState('')
  const state = useRouteState();
  useEffect(() => {
    // const {state} = window.history.state;
    setpageTitle(state.username)
    setvalue(Number(state.value) || 0)
    setuid(state.uid)
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
  return <div>
    <Navbar title={t(pageTitle)} />
    {uid&&<Tabs swipeable lineWidth={10} value={value} onChange={setvalue}>
      <Panel title={<span className={value === 0 ? 'active' : 'unactive'}>Following</span>}>
        <div className='user-follow-page'  style={{'--window-height':`${window.innerHeight}px`}}>
          <Following pageType={'following'} uid={uid}></Following>
        </div>
      </Panel>
      <Panel title={<span className={value === 1 ? 'active' : 'unactive'}>Followers</span>}>
        <div  className='user-follow-page'  style={{'--window-height':`${window.innerHeight}px`}}>
          <Following pageType={'fans'} uid={uid}></Following>
        </div>
      </Panel>
    </Tabs>}
  </div>
}
export default UserFollowPage