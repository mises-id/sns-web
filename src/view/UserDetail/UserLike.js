/*
 * @Author: lmk
 * @Date: 2022-01-05 11:26:57
 * @LastEditTime: 2022-01-05 15:49:43
 * @LastEditors: lmk
 * @Description: user post list
 */
import { likesData } from "@/api/status";
import PostItem from "@/components/PostItem";
import PullList from "@/components/PullList";
import { useChangePosts, useList } from "@/utils";
import React, { useEffect, useState } from "react";
import { useHistory } from "react-router-dom";
const UserPosts = (props) => {
  const history = useHistory()
  const [lastId, setlastId] = useState("");
  const [fetchData, last_id, dataSource, setdataSource] = useList(likesData, {
    uid:props.uid,
    limit: 20,
    last_id: lastId,
  });
  //getData
  useEffect(() => {
    setlastId(last_id);
  }, [last_id]);
  const { setLike } = useChangePosts(setdataSource, dataSource);
  const renderView = (val = {}, index) => {
    return (
      <PostItem
        val={val}
        key={index}
        index={index}
        history={history}
        setLike={setLike}
      />
    );
  };
  return (
    <div className="user-post-content" style={{'--window-height':`${window.innerHeight}px`}}>
      <PullList
        renderView={renderView}
        data={dataSource}
        load={fetchData}
      ></PullList>
    </div>
  );
};
export default UserPosts;
