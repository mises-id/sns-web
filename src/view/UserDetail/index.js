import React, { useEffect, useState } from "react";
import userBg from "@/images/user_bg.png";
import userBack from "@/images/user_back.png";
import userMore from "@/images/user_more.png";
import "./index.scss";
import Image from "@/components/Image";
import { useTranslation } from "react-i18next";
import MButton from "@/components/MButton";
import { objToUrl, urlToJson, useChangePosts, useLoginModal, username, useRouteState } from "@/utils";
import { useLogin } from "@/components/PostsIcons/common";
import addIcon from "@/images/add.png";
import { ActionSheet, Panel, Tabs, Toast } from "zarm";
import UserPosts from "./UserPost";
import UserLikes from "./UserLike";
import { useHistory } from "react-router-dom";
import { useDidRecover } from 'react-router-cache-route'
import { getUserInfo, JoinBlackList, removeBlackList } from "@/api/user";
const UserDetail = (props) => {
  const [userInfo, setUserInfo] = useState({});
  const { t } = useTranslation();
  const { isLogin } = useLogin();
  const [value, setvalue] = useState(0);
  const state = useRouteState();
  useEffect(() => {
    setInfo(state)
    setvalue(0)
    // eslint-disable-next-line
  }, [state.uid]); 
  useDidRecover(() => {
    const historyState = urlToJson(window.location.search);
    console.log(historyState)
    setInfo(historyState)
    if(JSON.stringify(historyState)!==JSON.stringify(state)){
      window.refreshByCacheKey("/userDetail")
    }
    window.refreshByCacheKey("/userFollowPage")
    window.refreshByCacheKey("/comment")
    window.refreshByCacheKey("/post")
  },[state])
  
  const setInfo = (state) =>{
    const uid = state.uid
    state.uid = ''
    setUserInfo(state);
    getUserInfo(uid).then(res=>{
      res.avatar = res.avatar ? res.avatar.medium : ''
      setUserInfo(res);
      const btnArr = res.is_blocked ? removeBlackButton : joinBlackButton;
      setbuttons(btnArr)
    })
  }
  const history = useHistory();
  // show assentSheet
  const [visible, setVisible] = useState(false)
  const joinBlackButton = [{ // Black user
    text:'Block',
    theme: 'danger',
    onClick: () => {
      blackUser()
      setVisible(false)
      Toast.show(t('succcess'))
    }
  }]
  const removeBlackButton = [{ // Black user
    text:'unBlock',
    theme: 'danger',
    onClick: () => {
      removeBlackUser()
      setVisible(false)
      Toast.show(t('succcess'))
    }
  }]
  // Judge the display content according to the status
  const [buttons, setbuttons] = useState(joinBlackButton)
  const blackUser = ()=>{
    JoinBlackList({uid:Number(state.uid)}).then(res=>{
      setbuttons(removeBlackButton)
    })
  }
  const removeBlackUser = ()=>{
    removeBlackList(state.uid).then(res=>{
      setbuttons(joinBlackButton)
    })
  }
  const showCheckItem = ()=> setVisible(true)
  const back = () => {
    window.history.back();
  };
  const loginModal = useLoginModal();
  const hasLogin = (e, fn) => {
    e.stopPropagation();
    if (!isLogin) {
      loginModal(fn, { type: "isLogin" });
      return false;
    }
    fn && fn();
  };
  const getFollow = (value) => {
    history.push({
      pathname: "/userFollowPage",
      search: objToUrl({ uid: userInfo.uid, username: username(userInfo), value }),
    });
  };
  const { followPress } = useChangePosts(setUserInfo, userInfo);
  const followedItem = (e) => hasLogin(e, async ()=>{
    await followPress(userInfo)
  });
  const isFollow = userInfo.is_followed ? "followedTxt" : "followTxt";
  return (
    <div>
      <div className="app-container">
        <div>
          {/* header */}
          <img src={userBg} className="user-bg" alt="userbg"></img>
          <div className="m-flex nav-box m-row-between">
            <img
              src={userBack}
              className="nav-icon"
              alt=""
              onClick={back}
            ></img>
            <img src={userMore} className="nav-icon" alt="" onClick={showCheckItem}></img>
          </div>
          {/* userinfo */}
          <div className="userinfo-box m-flex">
            <div className="avatar-box">
              <Image source={userInfo.avatar} size={75}></Image>
            </div>
            <div className="m-flex-1 user-info">
              <div className="m-flex m-row-between">
                <span className="username">{username(userInfo)}</span>
                <MButton
                  txt={t(isFollow)}
                  onPress={followedItem}
                  {...(userInfo.is_followed
                    ? {
                        borderColor: "#FFFFFF",
                        txtColor: "#FFFFFF",
                      }
                    : {
                        borderColor: "#FFFFFF",
                        backgroundColor: "#FFFFFF",
                        txtColor: "#666",
                      })}
                  imgIcon={!userInfo.is_followed ? addIcon : ""}
                  width={70}
                  height={25}
                />
              </div>
              <div className="follow-box m-flex">
                <div onClick={() => getFollow(0)}>
                  <span className="follow-num">{userInfo.followings_count || 0}</span>
                  <span className="follow-name">Following</span>
                </div>
                <div onClick={() => getFollow(1)}>
                  <span className="follow-num">{userInfo.fans_count || 0}</span>
                  <span className="follow-name">Followers</span>
                </div>
              </div>
            </div>
          </div>
        </div>
        {/* tab content */}
        {userInfo.uid && (
          <div className="tab-content">
            <div className="tab-header">
              <Tabs swipeable lineWidth={10} value={value} onChange={setvalue}>
                <Panel
                  title={
                    <span className={value === 0 ? "active" : "unactive"}>
                      Posts
                    </span>
                  }
                >
                  <UserPosts uid={userInfo.uid} />
                </Panel>
                <Panel
                  title={
                    <span className={value === 1 ? "active" : "unactive"}>
                      Likes
                    </span>
                  }
                >
                  <UserLikes  uid={userInfo.uid}/>
                </Panel>
              </Tabs>
            </div>
          </div>
        )}
      </div>
      <ActionSheet
        spacing
        visible={visible}
        actions={buttons}
        onMaskClick={() => setVisible(!visible)}
        onCancel={() => setVisible(!visible)}
      />
    </div>
  );
};
export default UserDetail;
