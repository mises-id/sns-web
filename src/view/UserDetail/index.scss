
.user-bg{
  width: 100%;
  position: absolute;
  left: 0;
  top: 0;
  z-index: 1;
}
.app-container{
  position: relative;
  // min-height: 100vh;
  // z-index: 11111;
  height: calc(100vh - var(--window-layout-height));
  overflow-y: auto;
}
.nav-icon{
  width: 30px;
  height: 30px;
}
@mixin headerBgView {
  position: relative;
  z-index: 2;
  padding-left: 15px;
  padding-right: 15px;
}
.nav-box{
  @include headerBgView;
  height: 60px;
}
.userinfo-box{
  @include headerBgView;
  padding: 20px 15px;
  box-sizing: border-box;
}
.avatar-box{
  border-radius: 50%;
  border: 2px solid white;
  width: 79px;
  height: 79px;
}
.user-info{
  margin-left:13px;
}
.username{
  font-size: 22px;
  color: white;
  font-weight: bold;
}
.follow-box{
  margin-top: 35px;
  font-size: 14px;
  .follow-name{
    color: #666;
  }
  .follow-num{
    color: #333;
    font-weight: bold;
    margin-right: 7px;
  }
  div{
    &:first-child{
      margin-right: 22px;
    }
  }
}
.active{
  font-size: 19px;
  font-weight: bolder;
  transition: all .2s;
  color: #333333;
}
.unactive{
  color: #666666;
  font-size: 15px;
}
.za-tabs{
  .za-tabs__header{
    background-color: white;
  }
}
.user-post-content{
  .m-layout{
    height: auto;
    min-height: 400px;
    overflow: hidden;
  }
}
.user-follow-page{
  .m-layout{
    height: calc(100vh - 90px - var(--window-layout-height));
  }
}
.user-detail-header{
  // background: url(~/static/images/user_bg.jpg) no-repeat;
  background-position: top;
  background-size: 100vw 115px;
  background-repeat: no-repeat;
  background-color: white;
}
.intro-content{
  background: #F8F8F8;
  border-radius: 5px;
  padding: 15px 26px;
  margin:  0 15px;
  position: relative;
}
.intro-txt{
  font-size: 14px;
  font-weight: 400;
  color: #333333;
  line-height: 19px;
}
.quotes{
  width: 8px;
  position: absolute;
  &.left{
    top: 11px;
    left: 7px;
  }
  &.right{
    bottom: 11px;
    right: 7px;
  }
}
.NFT-list{
  padding:0 15px 15px;
  .NFT-title{
    font-size: 19px;
    font-weight: bold;
    color: #333333;
    margin-top: 4px;
  }
}
.NFT-empty-line{
  height: 15px;
}
.nft-list-items{
  grid-template-columns: 1fr 1fr 1fr;
  gap: 15px;
  margin-top: 18px;
  .grid-item{
    img{
      width: 100%;
      height: calc((100vw - 60px) / 3);
    }
  }
}
.nft-name{
  margin-top: 7px;
  padding-right: 6px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-size: 12px;
  font-weight: 400;
  color: #333333;
  width: calc((100vw - 60px) / 3);
}
.view-more{
  width: 105px;
  margin: 5px auto 0;
  padding-bottom: 20px;
  span{
    font-size: 12px;
  }
  img{
    width:4px;
    margin-left: 5px;
  }
}
.tab-header{
  position: sticky;
  top: 45px;
  z-index: 100;
  background: white;
  .adm-tabs-tab-active{
    font-size: 19px;
    font-weight: bold;
    color: #333333;
  }
}
.tab-content{
  margin-top: 10px;
}
.adm-tabs{
  --active-line-height: 4px;
}