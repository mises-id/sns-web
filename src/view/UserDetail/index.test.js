/*
 * @Author: lmk
 * @Date: 2021-07-15 15:48:10
 * @LastEditTime: 2022-01-05 10:30:32
 * @LastEditors: lmk
 * @Description: UserDetail test
 */

import React from 'react';
import UserDetail from './';
import {shallow } from 'enzyme';
describe('UserDetail', () => {
  it('should render correctly', () => {
    const component = shallow(<UserDetail />)
    expect(component).toMatchSnapshot();
  });
});