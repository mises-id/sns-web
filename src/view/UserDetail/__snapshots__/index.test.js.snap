// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetail should render correctly 1`] = `
<div>
  <div
    className="app-container"
  >
    <div>
      <img
        alt="userbg"
        className="user-bg"
        src="user_bg.png"
      />
      <div
        className="m-flex nav-box m-row-between"
      >
        <img
          alt=""
          className="nav-icon"
          onClick={[Function]}
          src="user_back.png"
        />
        <img
          alt=""
          className="nav-icon"
          onClick={[Function]}
          src="user_more.png"
        />
      </div>
      <div
        className="userinfo-box m-flex"
      >
        <div
          className="avatar-box"
        >
          <Image
            size={75}
          />
        </div>
        <div
          className="m-flex-1 user-info"
        >
          <div
            className="m-flex m-row-between"
          >
            <span
              className="username"
            >
              Anonymous
            </span>
            <MButton
              backgroundColor="#FFFFFF"
              borderColor="#FFFFFF"
              height={25}
              imgIcon="add.png"
              onPress={[Function]}
              txt="followTxt"
              txtColor="#666"
              width={70}
            />
          </div>
          <div
            className="follow-box m-flex"
          >
            <div
              onClick={[Function]}
            >
              <span
                className="follow-num"
              >
                0
              </span>
              <span
                className="follow-name"
              >
                Following
              </span>
            </div>
            <div
              onClick={[Function]}
            >
              <span
                className="follow-num"
              >
                0
              </span>
              <span
                className="follow-name"
              >
                Followers
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <ForwardRef(forwardRef)
    actions={
      Array [
        Object {
          "onClick": [Function],
          "text": "Block",
          "theme": "danger",
        },
      ]
    }
    onCancel={[Function]}
    onMaskClick={[Function]}
    spacing={true}
    visible={false}
  />
</div>
`;
